Docs for ES|QL Views (#146559)
elastic/elasticsearch: docs/reference/query-languages/esql/_snippets/common/comparing_views_subqueries_fork.md

[Views](/reference/query-languages/esql/esql-views.md),
[subqueries](/reference/query-languages/esql/esql-subquery.md) and the
[`FORK`](/reference/query-languages/esql/commands/fork.md) command are related.
There are many similarities and differences between them.

### High level definitions

* `FORK` allows data coming from previous commands, like an initial `FROM index` command, to be processed in parallel in multiple different branches, each performing different commands on the same original data.

> * Subqueries maintain that parallel processing capability, but allow each branch to use a different source index, a different `FROM` command per branch.

Reviewer: ```suggestion
* Subqueries also enable parallel processing, but allow each branch to use a different source index with a different `FROM` command per branch.
```

> * Views are similar to subqueries, each view has its own `FROM` command, but views are hidden behind view definitions, acting as _virtual indexes_.

Reviewer: ```suggestion
Views are reusable, named queries that act like virtual indices. Each view has its own `FROM` command and processing pipeline, and can be referenced like a regular index.
```

nit for slightly clearer phrasing

### Similarities

> All three mechanisms will process the entire set of query definitions at query time, resulting in an up-to-date response when source indexes are changed and the query is re-run.

Reviewer: I'd consider continuing the bulleted list formatting for these H3s too, for readability, with a nice lead-in phrase for each list

Author: Done

Columns from the results of multiple branches are merged into the main query, expanding the table of results, and inserting `null` values if any branch has different columns than the others.
Complex processing commands can be used inside both views and subqueries, as detailed in the [description of subqueries](/reference/query-languages/esql/esql-subquery.md#description).

> In theory, none of these techniques are capable of supporting nested branching (ie. subqueries within subqueries, or `FORK` within subqueries, etc.), however views do support this to some extent, as detailed below.

Reviewer: > In theory, none of these techniques are capable of supporting nested branching ... however views do support this to some extent 

This reads like its contradicting itself and is also a bit vague:D  

Consider: "Nested branching is generally not supported, but views can work around this limitation through query compaction."

Author: I'll take your suggestion, and see if it is clear enough for others too.

All of these approaches to parallel processing are bound by the same maximum branch count of 8.

> ### Differences with FORK

Reviewer: ```suggestion
### FORK differences
```

Author: Done

The `FORK` command never includes a `FROM` command, and relies entirely on an existing query to provide the incoming columns.
This also means that all branches will receive identical incoming data, the same columns and the same rows.
This is not true of subqueries or views, which can receive completely different columns and rows from their own `FROM` commands.
Nested branches are not allowed, which means `FORK` branches cannot contain other `FORK` commands.

> In fact it is not allowed to use more than one `FORK` command in the same query.

Reviewer: ```suggestion
Only one `FORK` command is allowed per query, so nested branches are not possible.
```

the stronger constraint seems to make the original first sentence redundant so could simplify here I think

> This limitation is partially true for views and subqueries, but to a lesser extent as described below.

Reviewer: ```suggestion
Views and subqueries have similar restrictions, but views can work around some of them through query compaction.
``` 

"Partially true but to a lesser extent" is very careful hedging, reads vague 🤓

Author: Done
